diff -ru xen-4.1.1/tools/ioemu-qemu-xen/vl.c xen-4.1.1/tools/ioemu-qemu-xen/vl.c
--- xen-4.1.1/tools/ioemu-qemu-xen/vl.c	2011-02-11 18:54:51.000000000 +0100
+++ xen-4.1.1/tools/ioemu-qemu-xen/vl.c	2012-02-21 13:52:43.320000307 +0100
@@ -52,6 +52,10 @@
 
 #include <xen/hvm/hvm_info_table.h>
 
+#ifdef CONFIG_STUBDOM
+#include <qubes_gui_qemu.h>
+#endif
+
 #include <unistd.h>
 #include <fcntl.h>
 #include <signal.h>
@@ -5984,7 +5988,9 @@
         }
     } else { 
 #ifdef CONFIG_STUBDOM
-        if (xenfb_pv_display_init(ds) == 0) {
+        if (qubesgui_pv_display_init(ds) == 0) {
+            fprintf(stderr, "qubes gui initialized\n");
+        } else if (xenfb_pv_display_init(ds) == 0) {
         } else
 #endif
 #if defined(CONFIG_CURSES)
diff -ru xen-4.1.1/tools/ioemu-qemu-xen/xen-hooks.mak xen-4.1.1/tools/ioemu-qemu-xen/xen-hooks.mak
--- xen-4.1.1/tools/ioemu-qemu-xen/xen-hooks.mak	2011-02-11 18:54:51.000000000 +0100
+++ xen-4.1.1/tools/ioemu-qemu-xen/xen-hooks.mak	2012-02-21 13:51:46.908000310 +0100
@@ -45,6 +45,12 @@
 CONFIG_SDL=
 CONFIG_AUDIO=
 OBJS += xenfbfront.o
+
+OBJS+= qubes_gui.o txrx-vchan.o double_buffer.o qubes_gui_keymap.o
+LIBS+= -lvchan
+CPPFLAGS += -I$(XEN_ROOT)/tools/vchan -I$(XEN_ROOT)/tools/qubes-gui/include
+VPATH+=:$(XEN_ROOT)/tools/qubes-gui/gui-common:$(XEN_ROOT)/tools/qubes-gui/gui-agent-qemu
+
 else
 CPPFLAGS+= -I$(XEN_ROOT)/tools/blktap/lib
 LIBS += -L$(XEN_ROOT)/tools/blktap/lib -lblktap
